### ISSUE TITLE
Beta testing

### ISSUE THREAD

**Issue by Goli4thus** [Goli4thus/linvstmanager#1]
(takeover from https://github.com/Goli4thus/linvstmanage/issues/1)

--------------------------

First of all, thanks for your feedback! Definitely helps in getting this application to a more stable point!

> @Goli4thus
> First impressions just testing scan since i wonder how it will go:
> 
>     1. There are some strange things about build install, i'll describe a bit later...
> 
>     2. Scan of my VST folder of 2156 .dll files took just ~16 seconds that's pretty good! smiley
> 
>     3. There were quite a few excessive non-vst dlls on scan, but seems that mirrors pretty much what DAW scanners usually detect

The fact that it only took 16 seconds for 2156 .dll files makes me think that you probably didn't have the "Verify dll-files for being actual VST files" option ticked.
This requires setting up the _VstDllCheck.exe_ tool in preferences, which is shipped with _LinVstManager_.
It would also be the explanation why the scan showed you so many "non-vst dlls" as a result (VST2 files that weren't verified are indicated by 'VST2 (?)' in the scan result table).

> 
>     4. There's huge problem i have encountered, i've added all those scanned ones, and tried to **Enable** all those who are disabled this caused at the very least something like memory leak, but then DE even asked if i want to close the process because it's not responding (looking in my .so folder it stuck hard on 23 items). I thought that could be because i haven't yet saved config, so i have killed process, cleaned .so folder and started over, saved config, rebooted **linvstmanager** - tried again it goes same. Also i have noticed that to create 1 symlink it take ~4 seconds or even more (x 2000, you can imagine what a bummer that is), it's insanely much slow **Enable** action, compared to **linvstmanage** almost immediately creates all of my vst links..

I've spent lots of hours on this one today and came up with LOTS of performance improvements (28f47ce). 
I also checked with 2000 "dummy vst dlls" (empty dll-files) and was able to reproduce your findings with the slowness compared to _linvstmanage-cli_. Honestly don't know why I haven"t tested such big numbers myself beforehand. Anyway, please have another try. It should be noticably faster now.
    
> 
>     5. I have noticed that you write .so directly as **VstName.so**, which is i can tell from personal experience will be huge problem, because there are a lot of different devs who have exactly same vst names, therefore we must use at the very least something like **DevName_VstName.so,** like:
> 
> 
> ```
> Abbeyroadplugins_EMI_TG_12413_Limiter
> ```

I actually thought a bit about that and somewhat hoped that every Dev would try to give his VST a unique name to "stand out". But that probably was wishful thinking on my part and I get your point: It's not really robust like it is now and could easily cause clashing in the link folder.

Therefore I've introduced a suffix that is appended to soft-links in link folder (d2a43e6e467972ae38c9364d630502a4ee449952). This is based on a hash value that is calculated based on a VST's absolute path.
So unless two identically named VST.dll (or vst3) files are located within the same folder (which is simply not possible), that link name with the added suffix will always be unique.

> 
>     1. Import previous although works perfectly as intended, won't help me personally too much, because 4 and 5 also happens there, and prior i have followed convention of:
> 
> 
> ```
> ~/.vst/Abbeyroadplugins_EMI_TG_12413_Limiter/
>     TG12413_1969.so
>     TG12413_2005.so
> 
> ~/.vst/Aegean_music_Spirit_reverb/
>     Spirit Reverb x64.so
> ```
> 
> So unless it's followed my structure exactly (which is not possible since it's manual dev naming, sometimes it's not exactly same as what's embedded in vst as dev) then there's not too much sense, because i have customly configured Renoise VST cache (with categories and all that). So for myself any deviation from that structure would anyway require manual intervention with Renosie. It's my personal problem, just wanted to mention smile

Well, I get your point, but yeah, there's no real way to consider such a custom naming scheme.

> 
> Please notify @osxmidi that it's not ready for production / LinVst integration yet, 4 and 5 are really serious ones.

I don't think he assumes that _LinVstManager_ is already "rock solid" or "ready for shipping" because that would be quite uncommon for newly developed software.
I also intentionally have not made a release on github and haven't written about it on reaper forum due to that.

> 
> Also feel free to guide me through any useful debug info about 4 and 5 which i can provide, since it's pretty hard to emulate my amount of VST plugins, which could certainly be factor..

As mentioned above, empty dll dummy files worked surprisingly well for reproducing your findings.
Having your number of actually VSTs still will be helpful I think to catch those more extreme edge cases which the average user might never trigger on his own.

> 
> Will back with more testing soon and update on 1

>     1. On **build** (no expert at all, just wonder):
>        
>        * Tested on Manjaro Deepin lacks panel icon, launcher icon...Pretty much all variants of icons, except blurry one on alt-tab and small in system manager, seems that by default you install only **/usr/local/share/icons/hicolor/48x48/apps/linvstmanager.png**, but as far as i know Linux can use svg icons directly? Well, even if png, you can fill all those sizes or most important ones, i guess: 16x16 20x20 22x22 24x24 32x32 36x36 40x40 48x48 64x64 72x72 96x96 128x128 192x192 256x256 384x384 512x512. And why not default **/usr/share/icons**?
>        * **/usr/local/share/applications/linvstmanager.desktop** just wonder, why not **/usr/share/applications**, shouldn't it be default for any Linux distro?

As far as I know, "manually installed" applications (meaning without the distro's package manager, but some like make/make install) usually should go into those "local" variant system folders (/usr/local/* instead of /usr/*), to keep out of the way of where a distro's package manager installs to. 

Regarding the icons:
I hoped it was fine, because for me on Manajro with XFCE I get icons where I would expect them without be pixelated.
But there's no problem in making other resolutions of those icons. They are part of it now (078df09).

> 
>     2. I've cleaned everything and tested again without touching or interacting with computer at all, just to see tested **Enable** with Import of my config, which are 1905 real tested vsts. It took 2 hours and 15 minutes...
> 
> 
>     * Technically it have completed
> 
>     * It seems to me that now process of creating symlinks goes in a same process with GUI which leads to that leaks and caused full LinVstManager freeze, if so probably it's a good idea to create all routines like update / enable / disable etc inside new process to decouple it from GUI
> 
>     * It have created 106 additional .so symlinks inside my wineprefix, that wasn't there before in root of my VST folder here (no idea on what criteria or why):

I'd say please test again with the added performance improvements. Lot's of things didn't seem right with the previous version when it came to having lots of VSTs in the table.
In general I agree that heavy processing should not be done in the main GUI event loop.
That's for instance the reason why the scan operation is being done in it's own thread.

Right now (as mentioned above) the performance has significantly improved (at least for me; curious on your case).
So it might be considered "ok-ish" now and if one doesn't always perform each operation with thousands of VST at the same time, it might not really be noticeable the way it's implemented at the moment.
But I'll note it down as a possible improvement.

> 
> 
> ```
> ~/.PlayOnLinux/wineprefix/VST/drive_c/VST
> ```
> 
> I can give you names of those symlinks if you want...I see no system there
> 
>     * I have compared lists of .so symlinks created with **linvstmanage** (definitely correct one, since it matches vst exactly) and same imported list with **linvstmanager**, there are some mismatches and bug with **.** character parsing:
> 
> 
> linvstmanage 	linvstmanager
> Acid Rack 2.0.so 	Acid Rack 2.so
> AMB ElectraBass v1.2.so 	AMB ElectraBass v1.so
> Eiosis AirEQ 5.1.so 	Eiosis AirEQ 5.so
> NUGEN VisLM2_5.1 64bit.so 	NUGEN VisLM2_5.so
> NUGEN VisLM2_7.1 64bit.so 	NUGEN VisLM2_7.so
> Oresus_v1.2.so 	Oresus_v1.so
> PPG Wave 3.V.so 	PPG Wave 3.so
> Shred 1.06.so 	Shred 1.so
> SIDizer_v1.4.so 	SIDizer_v1.so
> SoloRackFX Multi InOut.x64.so 	SoloRackFX Multi InOut.so
> SoloRackFX.x64.so 	SoloRackFX.so
> SoloRack Multi InOut.x64.so 	SoloRack Multi InOut.so
> SoloRack.x64.so 	SoloRack.so
> Spire-1.1.so 	Spire-1.so
> 
> Some have collapsed to one because of that:
> linvstmanage 	linvstmanager
> arcDev.Blittr.so
> arcDev.Cyclotron.X2.so
> arcDev.Dubb.Box.RC1.so
> arcDev.ET-200.so
> arcDev.ET-301.so
> arcDev.Resomatic.so 	arcDev.so
> Mr. Alias Pro x64.so
> Mr. Tater head.so 	Mr.so
> MR.MUTILATOR.so
> MR.WAVESHAPER.so 	MR.so

Actually, there is a system. ;) 
The Qt API I've used gave me the filename up until the first period character, which then I added the ".so" extension to.
Simply didn't think of filenames having periods within their name.
Anyway, I've corrected and tested this as well (2d72f41).

**Comment by keybreak** [comment]
Well, we're in the same Linux boat, so my pleasure to help wherever i can! :smiley: 

> The fact that it only took 16 seconds for 2156 .dll files makes me think that you probably didn't have the "Verify dll-files for being actual VST files" option ticked.
This requires setting up the VstDllCheck.exe tool in preferences, which is shipped with LinVstManager.
It would also be the explanation why the scan showed you so many "non-vst dlls" as a result (VST2 files that weren't verified are indicated by 'VST2 (?)' in the scan result table).

Ahh....That's a little more complicated i think, but you're right.
Yes i do use only VST2, and of course i've set up **VstDllCheck.exe** in preferences and have "Verify dll-files for being actual VST files" which was by default anyway...

I've noticed that default **~/.wine** was "kinda" created (with just couple of folders) and it wasn't there. I think problem is the fact that my VST wineprefix is made with PlayOnLinux and i use non-system Wine build from **~/.PlayOnLinux/wine/linux-amd64/5.1-staging**, i have multiple good reasons to go that route, and looking on this 2 issues:
https://github.com/osxmidi/LinVst/issues/130
https://github.com/osxmidi/LinVst/issues/28
I bet that **VstDllCheck** just as **LinVST** itself now works only with system Wine and default Wineprefix and expects them...So it's more for **osxmidi** i guess, but i'm struggling to formulate the issue correctly.

> I've spent lots of hours on this one today and came up with LOTS of performance improvements (28f47ce).

Kudos! Will test soon and report back here :+1: 

> I actually thought a bit about that and somewhat hoped that every Dev would try to give his VST a unique name to "stand out". But that probably was wishful thinking on my part and I get your point: It's not really robust like it is now and could easily cause clashing in the link folder.

Well...Sometimes they're unknown or untitled, but most of times they at least try :laughing: 

> Therefore I've introduced a suffix that is appended to soft-links in link folder (d2a43e6). This is based on a hash value that is calculated based on a VST's absolute path.
So unless two identically named VST.dll (or vst3) files are located within the same folder (which is simply not possible), that link name with the added suffix will always be unique.

I will test that out, but thinking about it - this might be problematic as well:
- As i've mentioned above, for example i use custom categorization in Renoise database cache (otherwise without proper categories i'll be completely lost in this ocean of plugins), this means that whatever chosen naming scheme in final folder it would be, it better be at least somewhat stable, and if it's based on full-path hash...any change could be disaster

- In addition to point 1 - now my prefix is fully portable, i can zip and completely transfer it to another disk or another computer, that would be problematic since any move would mean hash change.

I still need to actually test both assumptions, maybe i'm wrong, but according to my previous knowledge that should be the case...So what i suggest is to hash this prefix by either full path starting from **drive_c** (this would mean that prefix can be portable) or even just parent folder name of vst plugin in question, what do you think?

>  I also intentionally have not made a release on github and haven't written about it on reaper forum due to that.

Ok cool! It's so good lookin' and well documented, i would assume it's meant to be ready! :smile: 

> to keep out of the way of where a distro's package manager installs to.

Gotcha, new skill learned.

>  But there's no problem in making other resolutions of those icons. They are part of it now (078df09).

Thx, that should fix problem with some of them blurry ones when they expected to be bigger, although it's my bad about mentioning there's no icon - there was none because i was stupid enough to forget about icon cache...Which sorted itself out on next reboot :face_with_head_bandage: 

> Actually, there is a system. ;)

Sorry, those above i've listed in tables of course had system with that dreaded dot character :laughing: , i've meant this one as another case:
There was some symlinks inside my wineprefix located at the root of my VST folder
```
~/.PlayOnLinux/wineprefix/VST/drive_c/VST
```
here are them, maybe you'll have an idea...:
```
++chebyshev.so
++compand.so
++decimate.so
++matrix.so
+morphfilter.so
++phasor.so
+spectralgate.so
AMT Amp Leveling.so
AMT Multi-Max.so
BC Axiom VST.so
BC Chorus 4 VST(Mono).so
BC Chorus 4 VST(Stereo).so
BC Destructor VST(Mono).so
BC Destructor VST.so
BC DP Meter Pro 4 VST(Mono).so
BC DP Meter Pro 4 VST(Srnd).so
BC DP Meter Pro 4 VST(Stereo).so
BC Dynamics 4 VST(Mono).so
BC Dynamics 4 VST(Stereo).so
BC Flanger 3 VST(Mono).so
BC Flanger 3 VST(Stereo).so
BC Free Amp VST.so
BC FreqAnalyst 2 VST(Mono).so
BC FreqAnalyst 2 VST(Stereo).so
BC FreqAnalyst Multi 2 VST.so
BC Gain 3 VST(Dual).so
BC Gain 3 VST(Mono).so
BC Gain 3 VST(Stereo).so
BC Late Replies VST.so
BC Liny EQ 5 VST(Dual).so
BC Liny EQ 5 VST(Mono).so
BC Liny EQ 5 VST(Stereo).so
BC MB-5 Dynamix 2 VST(Mono).so
BC MB-5 Dynamix 2 VST(Stereo).so
BC MB-7 Mixer 2 VST(Dual).so
BC MB-7 Mixer 2 VST(Mono).so
BC MB-7 Mixer 2 VST(Stereo).so
BC Oscilloscope Multi 2 VST.so
BC PatchWork Synth VST.so
BC PatchWork VST.so
BC Phaser 3 VST(Mono).so
BC Phaser 3 VST(Stereo).so
BC Plug'n Script Synth VST.so
BC Plug'n Script VST.so
BC Protector 2 VST.so
BC Re-Guitar VST.so
BC StereoScope Multi 2 VST.so
BC Triple EQ 4 VST(Dual).so
BC Triple EQ 4 VST(Mono).so
BC Triple EQ 4 VST(Stereo).so
Blue Cat FreqAnalystPro VST(Mono).so
Blue Cat FreqAnalystPro VST(Stereo).so
Blue Cat ParametrEQ3 VST(Mono).so
Blue Cat ParametrEQ3 VST(Stereo).so
Blue Cat Remote Control 16 VST.so
Blue Cat Remote Control 32 VST.so
Blue Cat Remote Control 64 VST.so
Blue Cat Stereo ParametrEQ3 VST.so
Blue Cat StereoScope Pro VST.so
Blue Cat Widening ParametrEQ3 VST.so
BlueVerb DRV-2080.so
British NEQ-1972.so
BT Analog Chorus CH2S-3.so
BT Analog Phaser APH2S-3.so
BT Compressor CP2S-3.so
BT Equalizer BX2S-3.so
BT Equalizer PEQ2B-3.so
BT Equalizer PEQ322-3.so
BT Equalizer PEQ5B-3.so
BT Limiter LM2S-3.so
BT Stereo Imager ST2S-3.so
BT Tempo Delay DL3D-3.so
COSMOS.so
E-3B Compressor.so
E-Channel.so
elysia alpha master.so
elysia alpha mix.so
elysia karacter master.so
elysia karacter mix.so
elysia mpressor sc.so
elysia mpressor.so
elysia museq master.so
elysia museq mix.so
elysia niveau filter.so
elysia nvelope.so
elysia phils cascade.so
E-Maximizer.so
E-RetroVox.so
E-TubeTape Warmer.so
Liquid Gate II.so
Liquid Mod II.so
Liquid Phase II.so
MAGNETIC II.so
Program EQP-4.so
PulseTec EQs.so
Studio Channel SC-226.so
Ultrafunk fxCompressor R3.so
Ultrafunk fxDelay R3.so
Ultrafunk fxEqualizer R3.so
Ultrafunk fxGate R3.so
Ultrafunk fxModulator R3.so
Ultrafunk fxMultiband R3.so
Ultrafunk fxPhase R3.so
Ultrafunk fxReverb R3.so
Ultrafunk fxSurround R3.so
Ultrafunk fxWahwah R3.so
```

Will get back with testing this update soon!

**Comment by keybreak** [comment]
There was some warning on build which wasn't there before update:
**make -j4**
```
[  7%] Building VstDllCheck utility...
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
```
Nothing serious?

___________

Right away i have suspicion about mismatch detection after legacy config import, there's something weird:
1. Since last test i haven't touched LinVST, so there shouldn't be missmatch, when i have opened my previous list everything was ok and reported as **Enabled**.

2. I've cleaned link folder, cleaned config and started over with importing same config from legacy.

3. Once it was done list i got have been reported as all mismatches

4. Simple reboot of program fixed this, now entries have been correctly stated as **Disabled**

____________

Creation time for all of my links is definitely improved, it took 1 blink of my eye, compared to previous 2 hours and 15 minutes...You know what, that was a damn good optimization!! :rofl: 

___________

Now .so links names produced by **linvstmanager** are identical to **linvstmanage** (apart from hash suffix), which is good  news :+1:

**Comment by Goli4thus** [comment]
> I've noticed that default **~/.wine** was "kinda" created (with just couple of folders) and it wasn't there. I think problem is the fact that my VST wineprefix is made with PlayOnLinux and i use non-system Wine build from **~/.PlayOnLinux/wine/linux-amd64/5.1-staging**, i have multiple good reasons to go that route, and looking on this 2 issues:
> [osxmidi/LinVst#130](https://github.com/osxmidi/LinVst/issues/130)
> [osxmidi/LinVst#28](https://github.com/osxmidi/LinVst/issues/28)
> I bet that **VstDllCheck** just as **LinVST** itself now works only with system Wine and default Wineprefix and expects them...So it's more for **osxmidi** i guess, but i'm struggling to formulate the issue correctly.
> 
Yeah, I'm not the wine expert here. As already mentioned, _VstDllCheck.exe_ is just a stripped down version of _TestVst.exe_ by osxmidi. So it's natural that if _TestVst.exe_ has issues in certain situations or simply behaves a certain way that doesn't play nicely with PlayOnLinux, here we have them as well. I also have no experience with PlayOnLinux what so ever. But it makes sense to me that _VstDllCheck.exe_ would trigger the creation of a default wine prefix, if there isn't one yet.

> > Therefore I've introduced a suffix that is appended to soft-links in link folder ([d2a43e6](https://github.com/Goli4thus/linvstmanager/commit/d2a43e6e467972ae38c9364d630502a4ee449952)). This is based on a hash value that is calculated based on a VST's absolute path.
> > So unless two identically named VST.dll (or vst3) files are located within the same folder (which is simply not possible), that link name with the added suffix will always be unique.
> 
> I will test that out, but thinking about it - this might be problematic as well:
> 
>     * As i've mentioned above, for example i use custom categorization in Renoise database cache (otherwise without proper categories i'll be completely lost in this ocean of plugins), this means that whatever chosen naming scheme in final folder it would be, it better be at least somewhat stable, and if it's based on full-path hash...any change could be disaster
> 
>     * In addition to point 1 - now my prefix is fully portable, i can zip and completely transfer it to another disk or another computer, that would be problematic since any move would mean hash change.
> 
> 
> I still need to actually test both assumptions, maybe i'm wrong, but according to my previous knowledge that should be the case...So what i suggest is to hash this prefix by either full path starting from **drive_c** (this would mean that prefix can be portable) or even just parent folder name of vst plugin in question, what do you think?
> 
Here's what my initial assumption was when implementing that change:
The name of the soft-link or even the name of the VST dll file itself don't really matter, because a DAW will read the actual VST name and vendor from the dll and display that in it's FX browser.
Therefore I thought I can easily add such a suffix and the user wouldn't even need to care about it. It would just be that thing in the hidden link folder.

But I just did some tests myself:
Reaper really works like that the suffix I add doesn't show up in Reaper's FX browser.
**BUT**:
Just downloaded the demo of Renoise and tested the same thing.
Renoise actually takes the name of the soft-link and displays that as the VST name in it's FX browser.
I actually did not thing that any DAW would behave like that. But I#ve been proven wrong now and totally agree that the introduced suffix is no real solution, because it simply looks bad in the DAW.

Regarding "make the hash bash on a path up until 'drive_c' or something":
There's no real guarantee that someone puts all his VSTs in wine prefixes.
It's perfectly possible to put you VSTs in a normal folder and apply LinVst there (it will just load whatever the last wine prefix happened to be). So it's no real solution.
Furthermore: Oftentimes VST go into the same folder in a wine prefix. Something like _"drive_c/Program Files/Steinberg/VSTPlugins/"_, which would once again introduce the possibility that two VSTs would get the same hash. That's why I thought taking the full path would work.

Regarding "moving the wine prefix would change the hash":
Well, moving the wine-prefix would cause the contained VSTs not being found anymore by LinVstManager (status _NotFound_) and would result in the user needing to setup those VSTs once again.

Just did another test:
I bridge the same VST with the same so-link name in two different folders and let Renoise scan both of them.
Result: Renoise only shows one of the two in its FX browser and ignores the other.

Reaper would simply detect that it in fact is the same VST and ignore the duplicate one.
Or if its two times "EQ.dll" from differnt vendors, Reaper would read those vendor strings and label them accordingly in its FX browser.

**Question**: How do you currently get Renoise to load two VSTs with the same name of the dll file? Do you rename one of them intenionally?

> > Actually, there is a system. ;)
> 
> Sorry, those above i've listed in tables of course had system with that dreaded dot character laughing , i've meant this one as another case:
> There was some symlinks inside my wineprefix located at the root of my VST folder
> 
> ```
> ~/.PlayOnLinux/wineprefix/VST/drive_c/VST
> ```
> 
> here are them, maybe you'll have an idea...:
> 
> ```
> ++chebyshev.so
> ++compand.so
> ...
> Ultrafunk fxWahwah R3.so
> ```
> 
> Will get back with testing this update soon!

I actualy don't really understand yet.
Are the VST.dll files located in `~/.PlayOnLinux/wineprefix/VST/drive_c/VST`?
Have you added them and performed an update in _LinVstManager_?
If so, it would totally make sense that those so-files are there.
But if they were real symlinks: Have you accidentely setup the linkfolder in preferences to that path at some point?


> There was some warning on build which wasn't there before update:
> **make -j4**
> 
> ```
> [  7%] Building VstDllCheck utility...
> make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> ```
> 
> Nothing serious?
Yeah I get that all the time as well.
Somehow the makefile of VstDllCheck causes that I get the impression.
But `make -j4` sure speeds up the build of _LinVstManager_ and overall it has no negative effect. That's why I wrote it in the README.

> 
> Right away i have suspicion about mismatch detection after legacy config import, there's something weird:
> 
>     1. Since last test i haven't touched LinVST, so there shouldn't be missmatch, when i have opened my previous list everything was ok and reported as **Enabled**.
> 
>     2. I've cleaned link folder, cleaned config and started over with importing same config from legacy.
> 
>     3. Once it was done list i got have been reported as all mismatches
> 
>     4. Simple reboot of program fixed this, now entries have been correctly stated as **Disabled**
> 
Strange indeed. Hard to diagnose though from that description alone.
Maybe observe if it happens again or was just a one off thing and if it happens again, try to find out how you can reproduce it.
Potential bugs that a developer can't reproduce himself are the among the hardest bugs to fix.

> 
> Creation time for all of my links is definitely improved, it took 1 blink of my eye, compared to previous 2 hours and 15 minutes...You know what, that was a damn good optimization!! rofl
> 
> Now .so links names produced by **linvstmanager** are identical to **linvstmanage** (apart from hash suffix), which is good news +1

Both good news! Glad that "performance update" made that much of a difference.

**Comment by keybreak** [comment]
> Yeah, I'm not the wine expert here.

Thx God for that, otherwise you would lay drunk somewhere :laughing: 
Problem is that it makes us 3...
Only people who actually know Wine are those who actually do it :worried: 

> I also have no experience with PlayOnLinux what so ever.

Just a wrapper to be honest (if you use it to create / manage / configure prefixes), i had took time to compare it in order to drop later - it's exactly same as Wine itself in terms of creation multiple prefixes, but they had a brilliant idea of having different pre-compiled Wine versions absolutely independent ow Wine / system - which makes a lot of sense, for example i've tested Wine 5.1 which just came out - it kills registration for few on plugins, if you back off to 5.0 - it's ok...So in order to keep really stable studio, especially in my case - it makes total sense to use this type of approach to wrok with Wine.

_______________


### VstDllCheck

> As already mentioned, VstDllCheck.exe is just a stripped down version of TestVst.exe by osxmidi. So it's natural that if TestVst.exe has issues in certain situations or simply behaves a certain way that doesn't play nicely with PlayOnLinux, here we have them as well. But it makes sense to me that VstDllCheck.exe would trigger the creation of a default wine prefix, if there isn't one yet.

Yeah...I guess it's matter of making few fixes in order to make LinVst / TestVst.exe / VstDllCheck.exe behave with certain environment variables, i need to think on how to formulate it, but best chances we got is with osxmidi...I'll link issue here once i'll get to right form.

So i did few more tests and at least for now i suggest to add in readme and / or program itself that in order to **actually** use **VstDllCheck.exe** you must have **~/.wine** created prior with Wine itself, because:
- **VstDllCheck.exe** can NOT create it itself, it will try but will create just those few folders:

**ls -lAR**
```
.:
total 8
drwxr-xr-x 2 x133 x133 4096 feb  8 21:21 dosdevices
drwxr-xr-x 3 x133 x133 4096 feb  8 21:21 drive_c

./dosdevices:
total 0
lrwxrwxrwx 1 x133 x133 10 feb  8 21:21 c: -> ../drive_c
lrwxrwxrwx 1 x133 x133  1 feb  8 21:21 z: -> /

./drive_c:
total 4
drwxr-xr-x 4 x133 x133 4096 feb  8 21:21 windows

./drive_c/windows:
total 8
drwxr-xr-x 2 x133 x133 4096 feb  8 21:21 system32
drwxr-xr-x 2 x133 x133 4096 feb  8 21:21 syswow64

./drive_c/windows/system32:
total 0

./drive_c/windows/syswow64:
total 0

```
- If you will use it just like that it won't do anything on scan, it will be just like without ****VstDllCheck.exe****

__________

### Mismatch detection

It's 100% reproducible for me like that on Manjaro Stable:
1. Emulate 1st launch by cleaning
**~/.vsttest** which is my link folder for tests
**~/.config/linvst** which is config
2. Launch program, configure it as always
3. Import legacy config - it will say mismatched by default for all entries
4. But if you save and reboot program - it will correctly report all as disabled

That must be just some problem with updating post legacy import status i believe, not a big deal really but may be little misleading if you don't know about it :wink: 

Current environments, if it helps:
```
Kernel: 5.4.17-1-MANJARO x86_64
compiler: gcc v: 9.2.0
tk: Qt 5.14.1
```

______________

I have a lot of valuable info to share on other issues, but need some sleep because otherwise will collapse, tomorrow i'll add up on other questions / issues :sleeping:

**Comment by keybreak** [comment]
### Symlinks appeared where they should not

This happened only once with initial version, i wasn't able to reproduce it yet, maybe it was because of crash.

> I actualy don't really understand yet.

Yeah forgive me, it's not obvious how my VST folder is constructed, sometimes i forget that...
That is my core folder:
```
~/.PlayOnLinux/wineprefix/VST/drive_c/VST
```
And it does **not** have any direct .dll files:
**ls -lA**
```
total 20
drwxr-xr-x 41 x133 x133 4096 oct 25 12:56  VST
drwxr-xr-x 99 x133 x133 4096 sep 24 16:15  VSTi
drwxr-xr-x 71 x133 x133 4096 aug 31 21:19 'VSTi NI'
drwxr-xr-x 27 x133 x133 4096 mar 31  2019 'VST NI'
```

So it is really weird that some was created, i believe they have been created first time when crash occurred, but again  it happened once and not with new version, at least yet.

> But if they were real symlinks: Have you accidentally setup the linkfolder in preferences to that path at some point?

I doubt it...But even if not, in theory that is possible!
Maybe you should create fool-proof check that would not allow to select **Link foler** in settings, if it's path have **/drive_c/**?

**Comment by keybreak** [comment]
### Symlinks naming

Problems:
- There are different DAWs (not only Renoise, at least some time ago Ableton Live / FL / AudioMulch) which can use actual file naming (at least as defaults) in some parts of UI

- Symlinks naming convention must be something as stable as **Aegean_music_Spirit_reverb/Spirit Reverb x64.so** and independent of any changes outside of given wine prefix (please see **Struggle of having boatload of plugins** for expanded info on why it is important)

_________________

> Therefore I thought I can easily add such a suffix and the user wouldn't even need to care about it. 

For the most part this is still true for most users which doesn't group / rename plugins at all...

> totally agree that the introduced suffix is no real solution, because it simply looks bad in the DAW.

Although even myself i see this as minor annoyance compared to rest of the problem - i'm really glad that you see this as well :+1: 

> There's no real guarantee that someone puts all his VSTs in wine prefixes.
It's perfectly possible to put you VSTs in a normal folder and apply LinVst there (it will just load whatever the last wine prefix happened to be). So it's no real solution.

Wait what? You can install windows stuff in Linux outside of Wineprefix?
This is a very bad idea, because it means you will:
- Loose portability of Wineprefix, which is super-important for serious work
- Can easily break vst plugin behavior for a lot of various reasons

So i still view this as possible stable-ish solution (outside of aesthetics).
And also i think we should discourage use of VSTs outside of actual Wine prefix...

> Regarding "moving the wine prefix would change the hash":
Well, moving the wine-prefix would cause the contained VSTs not being found anymore by LinVstManager (status NotFound) and would result in the user needing to setup those VSTs once again.

Sorry, again i forgot to mention few key details...
Of course i meant to go smart about it :smile:

Meaning that let's say your prefixes are located at
```
~/.PlayOnLinux/wineprefix
``` 
And if you decide to move your VST prefix to any location - you just symlink it to your prefixes location, this way you won't have not found and need to setup everything, because it's same path :sunglasses: 

Believe me, i've tried :+1: 
This is one of key reasons i wouldn't ever change Wine for anything including Windows itself, because once you setup 2000 plugins (which you can imagine is....a LOT of time) you won't ever need to do this again!! And even more, let's say you want to move all your plugins from studio to laptop - you can do this simply by copy whole wineprefix to laptop, without need to waste your time ever again, it won't even break registration of plugins!

Extremely powerful knowledge to have :wink:

> Furthermore: Oftentimes VST go into the same folder in a wine prefix. Something like "drive_c/Program Files/Steinberg/VSTPlugins/", which would once again introduce the possibility that two VSTs would get the same hash. That's why I thought taking the full path would work.

This case is one of those:
- Extremely rare, almost impossible case where installer of different plugins with same .dll name will at the same time write in core folder like **drive_c/Program Files/Steinberg/VSTPlugins/"** without creating new folder inside and overwrite previous .dll without asking or at least throwing warning
- Regardless of our decision here, even on Windows - if user writes all his vst plugins directly in same folder he's doomed in a long run, see why below.

So i still think this is best idea so far, if we'll use hash as a solution it should be based on path relative to **drive_c**...

> **Question**: How do you currently get Renoise to load two VSTs with the same name of the dll file? Do you rename one of them intentionally?

That is indeed a very good question, i have learned the right answer way back ago hard way.
Yes, sometimes you can rename actual .dll if it's free and/or not complicated ones which won't break - but there are a lot of cases with proprietary plugins when you can't rename .dll without consequences.

Therefore right answer is never put VSTs in same folder and use good separated structure!
Any other way - you're not guaranteed against problems.

```
drive_c/VST/VST/_TYPE_/_VENDOR_/_PLUGIN_/plugin.dll
drive_c/VST/VSTi/_VENDOR_/_PLUGIN_/plugin.dll
```

Actual example:
```
drive_c/VST/VST/EQ/Kush/Electra DSP/Electra DSPx64.dll
drive_c/VST/VSTi/u-he/ACE/ACE(x64).dll
```
________

### Renoise

It's not that simple actually. :thinking: 
At least i have double checked with my Renoise v3.2.1 (Linux x64)

1. Initially on scan list of VST and VSTi by default uses **DEV/NAME** (although sometimes dev is undefined / unknown) and in extremely rare cases when there's no NAME it uses just name of VST's .dll file

2. When you load VSTi it uses:
```
(VST): NAME (PRESET_NAME)
```

3. When you load VST
At the bottom panel:
```
(VST): NAME (->64bit)
```
In mixer view:
```
NAME
```

4. WM title of opened VST & VSTi
```
DEV: NAME (->64bit)
```

You can rename DEV / NAME of both VST and VSTi however you want and in all parts it will use new values, except WM title of opened plugin (it always uses values directly defined in plugin itself, not from Renoise db)

So i wonder, what's your Renoise demo version? What plugin you have used, was it real one or some dummy? Or that was exactly extremely rare case where there's no NAME value in plugin itself?

I would really like to know it, to double-test on clean VM, seems i can't really repeat it...
__________

### Struggle of having boatload of plugins

Here's heaven for such user, properly grouped, properly named & categorized plugins:

| VST | VSTi |
| --- | --- |
| ![DeepinScreenshot_dde-desktop_20200209142337](https://user-images.githubusercontent.com/47599989/74101687-a15ccd00-4b4d-11ea-8bd7-02d3b962d3ed.png) | ![DeepinScreenshot_dde-desktop_20200209142130](https://user-images.githubusercontent.com/47599989/74101697-b5083380-4b4d-11ea-9449-e0c6db61e449.png) |

To achieve this you would need to go one of 2 routes:
- Change one by one through properties inside Renoise
- Change db directly **~/.renoise/V3.2.0/CachedVSTs_x64.db**

**CachedVSTs_x64.db** can be opened with something like https://github.com/sqlitebrowser/sqlitebrowser and consists of 2 tables

**CachedLocalFileInfo** - this is db of files, there's no problem here for any changes or updates on our side - you can wipe the whole list and it will be recreated on next Renoise scan without affecting anyhow.

LocalFilePath | ModificationTime | CpuTypes
-- | -- | --
/home/x133/.vst/u-he_ACE/ACE(x64).so | 1574627327 | X86_64


**CachedPlugins** - this is db of entries with various settings

FileIdentifier | DocumentIdentifier | LocalFilePath | IsBridged | PresetPath | BankPath | IsSynth | IsVisible | IsFavorite | IsDpiAware | DisplayName | ShortDisplayName | FavoriteDisplayName | NeedsStaticProcessingBuffer | EnableAutoSuspending | SupportsMultiProcessors
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ACE(x64).so | ACE(x64) | /home/x133/.vst/u-he_ACE/ACE(x64).so | 0 |   |   | 1 | 1 | 1 | 0 | # MODULARS: u-he: Ace | Ace | Ace | 0 | 1 | 1

Now that is a huge problem if **LocalFilePath** changes in any way (like with full path hash suffix for example) - on next plugin scan in Renoise it will be assumed that this is completely different plugin and Renoise will create fresh entry without all those customizations, previous entry would still be there but since there are physically no more same file it's **LocalFilePath** will be empty and that means that only one with **LocalFilePath** will be used.

So imagine that such change would be made x2000 times it = complete disaster, and for me it's not an option.

Few cases to keep in mind:
1. Imagine that you have changed Linux user name (and God forbid forgot to backup your db), then scanned plugins with Renoise and all hashes now are outdated and have to match - this will create complete disaster and means that you would have to manually change suffix for all 2000 entries. You can argue that same thing would happen without suffix anyway, but that's not really the case, if only changed parameter is just username - you can run SQL query to replace it with new one in one command, with hashes it won't work like that since they're individual per plugin. This case could be solved with hashes relative to **drive_c/**

2. Same goes for changing location / renaming .so link folder

3. Let's say you have updated plugin and folder names / path / .dll name have changed (happens all the time) - that's easy to fix regardless of naming, since it's just one entry.

P.S. Just fun fact about how cool Renoise is in some aspects - not sure how other Programs solve this now (if they do at all, but last time i've checked none of them have this) let's say you had project with Amplitube 3 and now you have updated to Amplitube 4 (in most cases updates doesn't change API / automation mappings), but of course you'll be warned that Amplitube is missing...You just rename Renoise **.xrnx** to **.zip**, unzip it, change it's song.xml from Amplitube 3 to Amplitube 4, zip it back again to .xrnx - now you have exactly same working song with Amplitube 4 :heart_eyes:

**Comment by Goli4thus** [comment]
> So i did few more tests and at least for now i suggest to add in readme and / or program itself that in order to **actually** use **VstDllCheck.exe** you must have **~/.wine** created prior with Wine itself, because:
> ...
> 
>     * If you will use it just like that it won't do anything on scan, it will be just like without 
I always assumed that a wine appication would trigger the creation of a default wine prefix if there wasn't one yet to be found under `$WINE_PREFIX` (meaing `echo $WINEPREFIX` is empty and there is no `~/.wine/` folder to be found).

Furthermore I just tested this for myself (I'm also on manjaro stable btw; so we are even on the same page on that one):
Here's what I did:

1. Remove default `~/.wine` folder
2. Make sure `echo $WINEPREFIX` outputs nothing
3. Launch _LinVstManager_ and scan a folder containing dll files
- repeated this test with scaning both a "normal" folder outside of any wine prefix and additionally with a folder inside another wine prefix. ==>> Same result.

Result: A new wine prefix is created under `~/.wine/` (wine popup shows itself before the scan), _VstDllCheck.exe_  seems to work, the scan detects everything right.
Looking at the wine prefix afterwards with your command `ls -lAR` it gives me lots of output (much more than your example output above).
Really seems like a complete wine prefix.

So I can't reproduce this one. It seems to work just fine even if there isn't a preexisting wine prefix.
Btw: I'm running `wine-5.0 (Staging)`.

****VstDllCheck.exe****
> 
> 
> ### Mismatch detection
> 
> It's 100% reproducible for me like that on Manjaro Stable:
> 
>     1. Emulate 1st launch by cleaning
>        **~/.vsttest** which is my link folder for tests
>        **~/.config/linvst** which is config
> 
>     2. Launch program, configure it as always
> 
>     3. Import legacy config - it will say mismatched by default for all entries
> 
>     4. But if you save and reboot program - it will correctly report all as disabled
> 
> 
> That must be just some problem with updating post legacy import status i believe, not a big deal really but may be little misleading if you don't know about it wink

Alright, thanks for the explanation on how to reproduce it.
That should be fixed now (5807d21).
Btw: _VstDllCheck.exe_ currently is really only used during the scan itself. A mismatch detection is done without it.


> I have a lot of valuable info to share on other issues, but need some sleep because otherwise will collapse, tomorrow i'll add up on other questions / issues sleeping

Don't forget: We don't have a tight deadline or anything like that.
So no point in burning ourselves out within a couple days. ;)


> ### Symlinks appeared where they should not
> 
> This happened only once with initial version, i wasn't able to reproduce it yet, maybe it was because of crash.
> 
> > I actualy don't really understand yet.
> 
> Yeah forgive me, it's not obvious how my VST folder is constructed, sometimes i forget that...
> That is my core folder:
> 
> ```
> ~/.PlayOnLinux/wineprefix/VST/drive_c/VST
> ```
> 
> And it does **not** have any direct .dll files:
> **ls -lA**
> 
> ```
> total 20
> drwxr-xr-x 41 x133 x133 4096 oct 25 12:56  VST
> drwxr-xr-x 99 x133 x133 4096 sep 24 16:15  VSTi
> drwxr-xr-x 71 x133 x133 4096 aug 31 21:19 'VSTi NI'
> drwxr-xr-x 27 x133 x133 4096 mar 31  2019 'VST NI'
> ```
> 
> So it is really weird that some was created, i believe they have been created first time when crash occurred, but again it happened once and not with new version, at least yet.
> 
If it was just a one off thing with an old version (lots of changes were made these last couple days) then I'd say we wait until it happens again.


> > But if they were real symlinks: Have you accidentely setup the linkfolder in preferences to that path at some point?
> 
> I doubt it...But even if not, in theory that is possible!
> Maybe you should create fool-proof check that would not allow to select **Link foler** in settings, if it's path have **/drive_c/**?

Hmm, I don't a check like that would be good, because it would impose rules on where a link folder needs to be. Rules that might just be confusing for the average user. So I'd like to keep it how it is now: linkfolder can be a folder wherever you want.

-------------
Two more things:

## Duplicate VST.dll names

Regarding a question I had from above that probably got lost:

How do you currently get Renoise to load two VSTs with the same name of the dll file? Do you rename one of them intenionally?

My current standpoint on this:
I'll probably revert the change that introduced the suffix regarding the so-links in link folder.
Reason:
- if a DAW actually bases a VST's name on its filename (like Renoise does),:
    a) it's not nice for the user to have that random suffix show up in the DAW's fx browser
    b) it's not stable at all if one custom sorts his VST collection within the DAW, because moving and re-adding the same VST later on would change the suffix and mess up the custom sorting

Furthermore I think that the average user might seldomly run into this problem of having two identically named VST files.
And even if:
I think it's just up to the user to rename those VST.dll files to a unique name before adding it to _LinVstManager_ and consecutively to a DAW like Renoise.

It all would be easier if every DAW would handle it the way Reaper does. Then the current solution with a path based hash suffix for the so-link in the link folder would be the most robust and versatile solution.
But if there are DAW like Renoise, then I think it would even make sense to block adding duplicate named VSTs in _LinVstManager_ all together (currently you can add such duplicates, but I just realized that enabling them won't work, because two so-links with the same name can't exist in one link folder).
So the nicest thing for the end user in that case might be to inform him during _add_ that some VSTs have been skipped due to duplicate names. Probably best to mention their names and giving a hint to rename the actual dll files if possible and trying again.


## Suggestion
How about we try to create different issues for different problems from now on?
This issue already is getting out of hand regarding it's size with lots of different findings. It just makes it harder for everyone involved to keep up with what is a new bug, is being discussed, is already fixed.

**Comment by Goli4thus** [comment]
> ### Symlinks naming
>...
> Wait what? You can install windows stuff in Linux outside of Wineprefix?
> This is a very bad idea, because it means you will:
> 
>     * Loose portability of Wineprefix, which is super-important for serious work
> 
>     * Can easily break vst plugin behavior for a lot of various reasons
> 
> 
> So i still view this as possible stable-ish solution (outside of aesthetics).
> And also i think we should discourage use of VSTs outside of actual Wine prefix...

I'm just saying that wine allows. LinVst allows it as long as the VST itself can handle it.
Some VSTs need to be in a wineprefix while being loaded. Furthermore it's advisable to keep your VSTs in a wine prefix just so any dll-overrides or settings you might have made are being applied when the VST is loaded vis LinVst.
So yeah: in theory we could advice users to keep their VST.dll files with wine prefixes.

> 
> > Regarding "moving the wine prefix would change the hash":
> > Well, moving the wine-prefix would cause the contained VSTs not being found anymore by LinVstManager (status NotFound) and would result in the user needing to setup those VSTs once again.
> 
> Sorry, again i forgot to mention few key details...
> Of course i meant to go smart about it smile
> 
> Meaning that let's say your prefixes are located at
> 
> ```
> ~/.PlayOnLinux/wineprefix
> ```
> 
> And if you decide to move your VST prefix to any location - you just symlink it to your prefixes location, this way you won't have not found and need to setup everything, because it's same path sunglasses
> 
> Believe me, i've tried +1
> This is one of key reasons i wouldn't ever change Wine for anything including Windows itself, because once you setup 2000 plugins (which you can imagine is....a LOT of time) you won't ever need to do this again!! And even more, let's say you want to move all your plugins from studio to laptop - you can do this simply by copy whole wineprefix to laptop, without need to waste your time ever again, it won't even break registration of plugins!
> 
> Extremely powerful knowledge to have wink
> 

Well, if there's still a symlink that points to the new location, then yeah, _LinVstManager_ should probably be able to find it.

Btw:
VSTs are being tracked by their absolute path (as seen in the main table in _LinVstManager_). That means if you move your wine prefix to a different PC with a different user name, all VSTs will be marked as _NotFound_.
This could be "fixed" by editing the xml config file directly and search and replace the username in all paths.

> > Furthermore: Oftentimes VST go into the same folder in a wine prefix. Something like "drive_c/Program Files/Steinberg/VSTPlugins/", which would once again introduce the possibility that two VSTs would get the same hash. That's why I thought taking the full path would work.
> 
> This case is one of those:
> 
>     * Extremely rare, almost impossible case where installer of different plugins with same .dll name will at the same time write in core folder like **drive_c/Program Files/Steinberg/VSTPlugins/"** without creating new folder inside and overwrite previous .dll without asking or at least throwing warning

I actually experienced some vendors that do this.
For instance: https://www.tokyodawn.net/tdr-nova/
It installs to `$WINE_PREFIX/drive_c/Program Files/Steinberg/VSTPlugins/`

> 
>     * Regardless of our decision here, even on Windows - if user writes all his vst plugins directly in same folder he's doomed in a long run, see why below.
> 
> 
> So i still think this is best idea so far, if we'll use hash as a solution it should be based on path relative to **drive_c**...

This also only works if the the path up to "drive_c" is unique.
If you got two vendors who both install to `$WINE_PREFIX/drive_c/Program Files/Steinberg/VSTPlugins/`, then you could end up with two wine prefixes (assuming one creates a wine prefix per vendor) that both contain a identically name VST (i.e. EQ.dll) that both have the same relative path up to "drive_c".


Overall I'm starting to feel like fighting the wrong battle here.
We can't possible catch every edge case that even DAWs might not do themselves.

Overall it seems to me that:
a) adding anything to the so-link in the linkfolder that is in anyway derived from its source VST.dll's path is a bad idea, because the path might change and possibliy messes up custom sorting in a DAW

b) deriving any information based on the relative path within a wine prefix (up to "drive_c") isn't bulletproof either


So the only idea I have now is:
- no more path based suffix of any kind 
- still allow to add identically names VST.dll files to _LinVstManager_ (differentiation is based on each VST's absolute path re it being unique)
- possibly highlight duplicate VST names (in name column) and allow user to rename only the internal name of tracked VSTs.

Reason:
This "name" that is shown in _LinVstManager_ is derived from the base VST.dll filename.
But after that it is only ever used to:
- be displayed in the table
- be the name of the *.so link created in the link folder.
So if we allow the user to rename VSTs within _LinVstManager_ (not the actual source VST.dll), then a user could give a unique name himself and that new name would end up being the so-link in the link folder and therefore the name that a DAW picks up (if that DAW likes picking up the filename instead of the embedded name within the dll).

What do you think?
Would this be sufficient solution?

**Comment by keybreak** [comment]
> How about we try to create different issues for different problems from now on?
This issue already is getting out of hand regarding it's size with lots of different findings. It just makes it harder for everyone involved to keep up with what is a new bug, is being discussed, is already fixed.

Totally makes sense!
I'll create new issues for those who are not fixed, on answering to your last couple of posts.

> I'm also on manjaro stable btw

Are you on Manjaro forums as well, by chance?
Sometimes it would make sense to connect quickly on PM there, it's very handy.

> Don't forget: We don't have a tight deadline or anything like that.
So no point in burning ourselves out within a couple days. ;)

Wise words....Sometimes it's hard to remember that, since you want to solve everything NOW! :rocket:

**Comment by keybreak** [comment]
### Mismatch detection

>  That should be fixed now (5807d21).

Yes, solved!


### Symlinks appeared where they should not

> If it was just a one off thing with an old version (lots of changes were made these last couple days) then I'd say we wait until it happens again.

Agree, for now it's marked solved but i'll keep that in mind, hopefully we won't see that again!

> So I'd like to keep it how it is now: linkfolder can be a folder wherever you want.

Ok, maybe not a rule but at least warning in program notes or readme could be a good tone...
But i guess it's pretty obvious not to do that anyway :smile:

**Comment by keybreak** [comment]
> VSTs are being tracked by their absolute path (as seen in the main table in LinVstManager). That means if you move your wine prefix to a different PC with a different user name, all VSTs will be marked as NotFound.
This could be "fixed" by editing the xml config file directly and search and replace the username in all paths.

Yes, that's not a problem at least for me mass renaming of user via xml directly is not a problem :smiley:

**Comment by keybreak** [comment]
> I think it's just up to the user to rename those VST.dll files to a unique name before adding it to LinVstManager and consecutively to a DAW like Renoise.

On that note, please read section **Renoise** in here https://github.com/Goli4thus/linvstmanager/issues/1#issuecomment-583844083
I have described in there how Renoise works for me now, and ask few questions because i was unable to reproduce this case where your mixer have followed naming of .so

_________

I believe every issue mentioned here was solved apart from couple of created ones :+1:

**Comment by Goli4thus** [comment]
> Are you on Manjaro forums as well, by chance?
> Sometimes it would make sense to connect quickly on PM there, it's very handy.

No, only every reading the "stable update news". But not really active there.


> I believe every issue mentioned here was solved apart from couple of created ones +1
Alright, then let's close this one up and continue in dedicated issues.

**Comment by keybreak** [comment]
@Goli4thus 
Here's attempt to formulate binaries problem for LinVst portability i was talking about, if you're interested:
https://github.com/osxmidi/LinVst/issues/143
